### PR TITLE
Set TOC depth freely for every note by using YAML metadata or an option within `[toc]`

### PR DESCRIPTION
**Pull request by zergar**
This commit allows one to set the depth of the table of contents freely by means of YAML metadata or an option directly within the `[toc]` tag (e.g. `[toc maxLevel=4]`. The default level can also be set by means of the environment variable `CMD_DEFAULT_TOC_DEPTH`, this variable defaults to 3. The toc depth of the automatically generated toc can only be set by setting the level in the YAML or below.

The precedence of the toc depth is as follows (highest to lowest): `[toc maxLevel=4]` > YAML metadata > `CMD_DEFAULT_TOC_DEPTH` > 3 (default value)

codimd-toc-example-customtoc

## Optional Features
+ [x] It might be nice to set a default depth with an environment variable, though I'm unsure how to access them from the frontend (as I found no example)

### CODE REVIEW

**Comment by zergar** [comment]
could resolve #368

**Comment by Yukaii** [comment]
> It might be nice to set a default depth with an environment variable, though I'm unsure how to access them from the frontend (as I found no example)

For now, you can take a look at PR 1306 to see how it pass environment variable config to frontend: https://github.com/hackmdio/codimd/pull/1306

**Comment by zergar** [comment]
> For now, you can take a look at PR 1306 to see how it pass environment variable config to frontend: #1306

Thank you very much for the tip, I've added the feature accordingly.

**Comment by zergar** [comment]
to finish this up, I've added the option to set a custom TOC level within the toc tag, e.g. `[toc maxLevel=4]`. I've updated the PR description accordingly.

File: lib/config/environment.js [hunk 1]
```diff
@@ -147,5 +147,6 @@ module.exports = {
   openID: toBooleanConfig(process.env.CMD_OPENID),
   defaultUseHardbreak: toBooleanConfig(process.env.CMD_DEFAULT_USE_HARD_BREAK),
   linkifyHeaderStyle: process.env.CMD_LINKIFY_HEADER_STYLE,
-  autoVersionCheck: toBooleanConfig(process.env.CMD_AUTO_VERSION_CHECK)
+  autoVersionCheck: toBooleanConfig(process.env.CMD_AUTO_VERSION_CHECK),
+  defaultTocDepth: toIntegerConfig(process.env.CMD_DEFAULT_TOC_DEPTH, 3)
```
Comment (Yukaii) [inline lib/config/environment.js:151]: The second argument is unnecessary.

```suggestion
  defaultTocDepth: toIntegerConfig(process.env.CMD_DEFAULT_TOC_DEPTH)
```

Comment (zergar) [reply lib/config/environment.js:151]: I'm approving this, but I'm not sure whether the toc depth isn't set anywhere else. A quick test though showed that somewhere a depth of 3 seems to be set.

**Comment by zergar** [comment]
@jackycute does this look proper to you?

**Pull request merged**